tempesta_fw/client.c

```diff
@@ -28,121 +28,136 @@
 #include "connection.h"
 #include "log.h"
 #include "procfs.h"
+#include "tdb.h"
+#include "lib/str.h"
+#include "lib/common.h"
 
-#define CLI_HASH_BITS	17
-#define CLI_HASH_SZ	(1 << CLI_HASH_BITS)
+static struct {
+	unsigned int db_size;
+	const char *db_path;
+	unsigned int lifetime;
+} client_cfg __read_mostly;
 
 typedef struct {
-	struct hlist_head	list;
-	spinlock_t		lock;
-} CliHashBucket;
+	TfwClient	cli;
+	time_t		expires;
+} TfwClientEntry;
 
-/*
- * TODO probably not the best container for the task and
- * HTrie should be used instead.
- */
-CliHashBucket cli_hash[CLI_HASH_SZ] = {
-	[0 ... (CLI_HASH_SZ - 1)] = {
-		HLIST_HEAD_INIT,
-	}
-};
-
-static struct kmem_cache *cli_cache;
-
-/* Total number of created clients. */
-static atomic64_t act_cli_n = ATOMIC64_INIT(0);
+static TDB *client_db;
 
 /**
  * Called when a client socket is closed.
  */
 void
 tfw_client_put(TfwClient *cli)
 {
+	TfwClientEntry *ent;
+	TdbRec *rec;
+
 	TFW_DBG2("put client %p, conn_users=%d\n",
 		 cli, atomic_read(&cli->conn_users));
 
 	if (!atomic_dec_and_test(&cli->conn_users))
 		return;
 
-	spin_lock(cli->hb_lock);
+	rec = (TdbRec *)((void *)cli - offsetof(TdbRec, data));
+	tdb_rec_get_lock(rec);
```

Comment:
`tfw_client_put()` releases the bucket read lock acquired in `tfw_client_obtain()`, so this is a nested and wrong lock.

Reply:
We're in big trouble actually with the solution and whole #1115: TDB is simply wasn't designed to keep directly referenced data. The current case for filter or cache tables is to lookup an entry, do some quick operations with it and forget about it. Clients handling is very different: we keep a pointer to stored TDB entry in many places as `TfwClient *peer`, i.e. once created, a client must have the same memory address.

Currenty TDB uses the read lock to protect a referenced entry from being reallocated on HTrie index split (node bursting), so while we use `TfwClient`, i.e. there is a `peer` pointer, the bucket can not be split. Having that we start from only 16 items in root node, we have plenty of splits first time and deadlocks will occur.

A possible solution could be dereference HTrie index each time on `peer` dereferencing, but it's too expensive. TDB must be extended to support small records with constant memory address. Actually, large records are always placed in separate buckets and only small records are compacted into one bucket. Let's leave making TDB nice for #515 and now do a very simple thing: make `TfwClient` data structure large enough to not to be processed in small records TDB work flow. Please don't forget to write `TODO #515` as a comment for the `TfwClient` data structure enhancement.